samza-core/src/main/java/org/apache/samza/metrics/reporter/MetricsSnapshotReporter.java

```diff
@@ -0,0 +1,228 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.ShellCommandConfig;
+import org.apache.samza.metrics.Counter;
+import org.apache.samza.metrics.Gauge;
+import org.apache.samza.metrics.MetricsReporter;
+import org.apache.samza.metrics.MetricsVisitor;
+import org.apache.samza.metrics.ReadableMetricsRegistry;
+import org.apache.samza.metrics.RegistryWithSource;
+import org.apache.samza.metrics.Timer;
+import org.apache.samza.serializers.Serializer;
+import org.apache.samza.system.OutgoingMessageEnvelope;
+import org.apache.samza.system.SystemProducer;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.util.Clock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * MetricsSnapshotReporter is a generic metrics reporter that sends metrics to a stream.
+ *
+ * jobName // my-samza-job
+ * jobId // an id that differentiates multiple executions of the same job
+ * taskName // container_567890
+ * host // eat1-app128.gird
+ * version // 0.0.1
+ * blacklist // Regex of metrics to ignore when flushing
+ */
+public class MetricsSnapshotReporter implements MetricsReporter, Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsSnapshotReporter.class);
+
+  private final SystemProducer producer;
+  private final SystemStream out;
+  private final int reportingInterval;
+  private final String jobName;
+  private final String jobId;
+  private final String containerName;
+  private final String version;
+  private final String samzaVersion;
+  private final String host;
+  private final Serializer<MetricsSnapshot> serializer;
+  private final Optional<String> blacklist;
+  private final Clock clock;
+
+  private final String execEnvironmentContainerId;
+  private final ScheduledExecutorService executor;
+  private final long resetTime;
+  private final List<RegistryWithSource> registries = new ArrayList<>();
+  private final Set<String> blacklistedMetrics = new HashSet<>();
+
+  public MetricsSnapshotReporter(SystemProducer producer, SystemStream out, int reportingInterval, String jobName,
+      String jobId, String containerName, String version, String samzaVersion, String host,
+      Serializer<MetricsSnapshot> serializer, Optional<String> blacklist, Clock clock) {
+    this.producer = producer;
+    this.out = out;
+    this.reportingInterval = reportingInterval;
+    this.jobName = jobName;
+    this.jobId = jobId;
+    this.containerName = containerName;
+    this.version = version;
+    this.samzaVersion = samzaVersion;
+    this.host = host;
+    this.serializer = serializer;
+    this.blacklist = blacklist;
+    this.clock = clock;
```

Comment:
`Clock` is nice to inject because it makes testing easier (i.e. mock the clock instead of relying on actual system timing). It might not be used in the existing test for this class, but it's a pattern I like to follow in general.